### PR TITLE
Formula interface

### PR DESCRIPTION
**Pull request by MartinStancsicsQC**
Checklist
* [x] Added a `CHANGELOG.rst` entry

## Summary

This PR adds support for specifying models via Wilkinson-formulas (just like in R or statsmodels). The implementation is based on [formulaic](https://github.com/matthewwardrop/formulaic), and is relatively full-featured (supports most of what statsmodels does). As a not-so-side effect, it allows for the on-the-fly specification of potentially complex interaction terms (c.f. #515 and #583).

## Related `tabmat` changes

Most of the machinery needed to make this possible is implemented in `tabmat`. Please also take a look at the related tabmat [PR #267](https://github.com/Quantco/tabmat/pull/267).

## New dependencies

This PR introduces a set of new dependencies. `glum` will directly depend on `formulaic`, which in turn requires `astor`, `cached_property`, `graphlib-backport`, `interface_meta`, and `wrapt`. Apart from `wrapt`, all of these are pure python packages, and even `wrapt` has a Python fallback.

## Example

The PR also includes a tutorial on how the formula interface works in `glum`. You can find the [notebook here](https://github.com/Quantco/glum/blob/formula/docs/tutorials/formula_interface/formula_interface.ipynb). To learn more about the formula interface on the `tabmat` side, please check out [this example notebook](https://gist.github.com/MartinStancsicsQC/eaad758d27cb119336804a3d4170aad1).

### CODE REVIEW

File: src/glum/_glm.py [hunk 1]
```diff
@@ -226,6 +228,53 @@ def _name_categorical_variables(
     return new_names
 
 
+def _parse_formula(
+    formula: Union[str, Formula], include_intercept: bool = True
```
Comment (matthewwardrop) [inline src/glum/_glm.py:232]: You could use `from formulaic import FormulaSpec` and `def _parse_formula(formula: FormulaSpec, ...)` here for greater generality :).

**Review by matthewwardrop — COMMENTED** [review]
Excited to see Formulaic gaining organic adoptions like this :).

File: src/glum/_glm.py (hunk 1 above)
Comment (MartinStancsicsQC) [reply src/glum/_glm.py:232]: Very cool, thanks for the tip!

File: tests/glm/test_glm.py [hunk 2]
```diff
@@ -2260,6 +2268,50 @@ def test_store_covariance_matrix(
...
 
 
+@pytest.mark.parametrize(
+    "formula", ["y ~ col_1 + col_2", "col_1 + col_2"], ids=["two-sided", "one-sided"]
+)
+def test_store_covariance_matrix_formula(regression_data, formula):
+    X, y = regression_data
+    df = pd.DataFrame(X, columns=[f"col_{i}" for i in range(X.shape[1])])
+
+    if "~" in formula:
```
Comment (MatthiasSchmidtblaicherQC) [inline tests/glm/test_glm.py:2278]: always the case in this test, right?

**Review by MatthiasSchmidtblaicherQC — CHANGES_REQUESTED** [review]

File: docs/tutorials/formula_interface/formula_interface.ipynb
[the remark this answers is not in the document]
Comment (MartinStancsicsQC) [reply docs/tutorials/formula_interface/formula_interface.ipynb:1]: Thanks Matthias, great ideas! Would it be okay for you if I put these suggestions on a "to be done before v3 release" backlog, but merge the PR before it's fully done so that we can test formulas in the pre-release?

File: src/glum/_glm.py [hunk 3]
```diff
@@ -232,6 +234,53 @@ def _name_categorical_variables(
...
+        The formula to parse.
+    include_intercept: bool, default True
+        Whether to include an intercept column if the formula does not
+        include (``+ 1``) or exclude (``+ 0`` or ``- 1``) it explicitly.
+
+    Returns
+    -------
+    tuple[Formula, Formula]
+        The left-hand side of the formula, the right-hand side of the formula,
+        and a boolean flag indicating whether or not an intercept should be
```
[the remark this answers is not in the document]
Comment (MartinStancsicsQC) [reply src/glum/_glm.py:260]: Oh, thanks, old behavior 😅

File: src/glum/_glm.py [hunk 4]
```diff
@@ -1847,50 +1912,109 @@ def _set_up_and_check_fit_args(
...
+                    )
+
+                    self.y_model_spec_ = y.model_spec
+                    y = y.A.ravel()
+
+                X = tm.from_formula(
+                    formula=rhs,
+                    data=X,
+                    include_intercept=False,
+                    ensure_full_rank=self.drop_first,
```
[the remark this answers is not in the document]
Comment (MartinStancsicsQC) [reply src/glum/_glm.py:1941]: Agreed. I'll expand the docstrings

File: tests/glm/test_glm.py (hunk 2 above)
Comment (MartinStancsicsQC) [reply tests/glm/test_glm.py:2278]: No, there are two test cases and the formula in the second one is simply `"col_1 + col_2"`. Formatting is a bit wonky though and it's hard to parse, but I blame black for that 😄

**Review by MatthiasSchmidtblaicherQC — APPROVED** [review]

File: docs/tutorials/formula_interface/formula_interface.ipynb
Comment (MatthiasSchmidtblaicherQC) [inline docs/tutorials/formula_interface/formula_interface.ipynb:1]: - Why not move the sneak peek before the table of contents?
- typo "forluma"
- "statsmodels/linearmodels" -> "statsmodels or linearmodels"
- In "The predictors should include the interaction~~s~~ of the categorical variables DrivAge and VehPower, as well as those two variables themselves.", add something like: "Neither the individual categoricals nor their interaction will be dummy-encoded by glum. For categoricals with many levels, this can lead to a substantial performance improvement over dummy encoding.

**Review by MatthiasSchmidtblaicherQC — COMMENTED** [review]
Some small suggestions. Otherwise, it looks good!

File: docs/tutorials/formula_interface/formula_interface.ipynb [hunk 5]
```diff
@@ -0,0 +1,1724 @@
+{
+ "cells": [
```
Comment (MarcAntoineSchmidtQC) [inline docs/tutorials/formula_interface/formula_interface.ipynb:2]: I agree with Matthias that the sneak peek would be better placed before the table of contents.

There is a warning that appears twice related to the intercept. If they are there on purpose (you want to show that a warning will appear), then do it just once. Otherwise, fix the warning.

Comment (MarcAntoineSchmidtQC) [reply docs/tutorials/formula_interface/formula_interface.ipynb:2]: Ohh and also, don't refer to "tutorial 1", use a more explicit name to the tutorial you are referring to.

**Review by MarcAntoineSchmidtQC — APPROVED** [review]
Great work! Well-tested PR and a super valuable addition to our features.

Please take a look at my small cosmetic comments.
